flight/probability/plot_algo.py

```diff
@@ -0,0 +1,424 @@
+        """
+        closest = None
+        closest_dist = float("inf")
+        for point in points:
+            dist = abs(point[0] - px[0]) + abs(point[1] - px[1])
+            if dist < closest_dist:
+                closest_dist = dist
+                closest = point
+        return closest
+
+    def in_corner(
+        self, pos: Tuple[int, int], history: list[Tuple[int, int]]
+    ) -> List[Tuple[int, int]]:
+        """
+        function to get the searcher out of a corner to the closest
+        undiscovered cell
+
+        Parameters
+        ----------
+        pos : Tuple[int, int]
+            the current position
+        history : List[Tuple[int, int]]
+            list of all visited points
+
+        Returns
+        -------
+        escape_path : List[Tuple[int, int]]
+            the path to reach the nearest unexplored cell
+        """
+
+        approx_nearest = self.find_closest(self.find_unseens(history), pos)
+        start = self.a_star_grid.node(pos[1], pos[0])
+        end = self.a_star_grid.node(approx_nearest[1], approx_nearest[0])
+        self.a_star_grid.cleanup()
+        path, _ = self.a_star.find_path(start, end, self.a_star_grid)
+
+        return [(x[1], x[0]) for x in path[1:]]
+
+    def breadth_search(self, start: Tuple[int, int]) -> List[Tuple[int, int]]:
+        """
+        A breadth based search algorithm to find the shortest path
+        to visit all cells
+
+        Parameters
+        ----------
+        start: Tuple[int, int]
+            the starting position, in terms of the compressed cell
+
+        Returns
+        -------
+        shortest_path : List[Tuple[int, int]]
+            A list of points that defines the shortest continous path
+            that visits all points in the compressed grid.
+        """
+        histories = [[start]]
+        while len(histories) != 0:
+            history = histories.pop(0)
+            if len(history) >= self.n:
+                if self.valid_solution(history):
+                    return history
+
+            possible_moves = self.get_valid_positions(history)
+            if len(possible_moves) == 0:
+                history += self.in_corner(history[-1], history)
+                possible_moves = self.get_valid_positions(history)
+
+            for move in possible_moves:
+                new_history = list(history)
+                new_history.append(move)
+                insort(histories, new_history, key=lambda x: len(x))
+
+
+class Decompressor:
+    @staticmethod
+    def __prep_grid(cell_map: CellMap):
+        new_grid = zeros((len(cell_map.data), len(cell_map[0])))
+        for i in range(len(cell_map.data)):
+            for j in range(len(cell_map[0])):
+                if cell_map[i][j].is_valid:
+                    new_grid[i][j] = 1
+        return new_grid
+
+    @staticmethod
+    def get_valid_point(point, cell_map, cell_size):
+        cell_map.display()
+        point = (point[0] * cell_size, point[1] * cell_size)
```

Comment:
Usually it's not good practice to redefine a parameter, so unless its absolutely necessary, make a new variable for this.